<?xml version="1.0" encoding="UTF-8"?>
<robot xmlns:xacro="http://wiki.ros.org/xacro">

    <!-- 因为只需要一台，不需要前缀-->
    <xacro:macro name="er10" 
                 params="joint_limited
                         link_1_lower_limit:=-2.8798 link_1_upper_limit:=2.8798
                         link_2_lower_limit:=-2.3562 link_2_upper_limit:=1.3963
                         link_3_lower_limit:=-1.3090 link_3_upper_limit:=2.8449
                         link_4_lower_limit:=-3.1416 link_4_upper_limit:=3.1416
                         link_5_lower_limit:=-2.2689 link_5_upper_limit:=2.2689
                         link_6_lower_limit:=-6.2832 link_6_upper_limit:=6.2832">
        <!-- 定义质量,SW中没有定义材质，值存在问题-->
        <xacro:property name="base_link_mass" value="23.987"/>
        <xacro:property name="link_1_mass" value="15.617"/>
        <xacro:property name="link_2_mass" value="2.9708"/>
        <xacro:property name="link_3_mass" value="11.338"/>
        <xacro:property name="link_4_mass" value="8.7242"/>
        <xacro:property name="link_5_mass" value="0.86404"/>
        <xacro:property name="link_6_mass" value="0.065514"/>

        <link name="base_link">
            <inertial>
              <origin xyz="-0.0056212 0 0.10746"
                      rpy="0 0 0" />
              <mass value="${base_link_mass}" />
              <inertia ixx="0" ixy="0" ixz="0"
                       iyy="0" iyz="0" izz="0" />
            </inertial>
        
            <visual>
              <geometry>
                <mesh filename="package://multibot_efort/meshes/er10/base_link.STL" />
              </geometry>
              <material name="Orange" />
            </visual>
        
            <collision>
              <geometry>
                <mesh filename="package://multibot_efort/meshes/er10/base_link.STL" />
              </geometry>
            </collision>
        </link>
        <!-- link1-->
        <link name="link_1">
          <inertial>
            <origin xyz="0.094373 -0.0050175 0.14722"
                    rpy="0 0 0" />
            <mass value="${link_1_mass}" />
            <inertia ixx="0" ixy="0" ixz="0"
                      iyy="0" iyz="0" izz="0" />
          </inertial>
      
          <visual>
            <geometry>
              <mesh
                filename="package://multibot_efort/meshes/er10/link_1.STL" />
            </geometry>
            <material name="Orange" />
          </visual>
      
          <collision>
            <geometry>
              <mesh
                filename="package://multibot_efort/meshes/er10/link_1.STL" />
            </geometry>
          </collision>
        </link>
        
        <joint name="joint_1" type="revolute">
          <origin xyz="0 0 0.21"
                  rpy="0 0 0" />
          <parent link="base_link" />
          <child link="link_1" />
          <axis xyz="0 0 1" />
          <xacro:unless value="${joint_limited}">
              <limit lower="${-2.0 * pi}" upper="${2.0 *pi}" effort="20" velocity="2.9671" />
          </xacro:unless>
          <xacro:if value="${joint_limited}">
              <limit lower="${link_1_lower_limit}" upper="${link_1_upper_limit}" effort="20" velocity="2.9671" />
          </xacro:if>
        </joint>
        <!-- link2-->
        <link name="link_2">
          <inertial>
            <origin xyz="0.0068855 -0.15544 0.30672"
                    rpy="0 0 0" />
            <mass value="${link_2_mass}" />
            <inertia ixx="0" ixy="0" ixz="0"
                      iyy="0" iyz="0" izz="0" />
          </inertial>
      
          <visual>
            <geometry>
              <mesh
                filename="package://multibot_efort/meshes/er10/link_2.STL" />
            </geometry>
            <material name="Orange" />
          </visual>
      
          <collision>
            <geometry>
              <mesh filename="package://multibot_efort/meshes/er10/link_2.STL" />
            </geometry>
          </collision>
        </link>
        
        <joint  name="joint_2" type="revolute">
          <origin  xyz="0.195 0 0.212"
                    rpy="0 0 0" />
          <parent link="link_1" />
          <child  link="link_2" />
          <axis xyz="0 -1 0" />
          <xacro:unless value="${joint_limited}">
              <limit lower="${-2.0 * pi}" upper="${2.0 *pi}" effort="20" velocity="2.9671" />
          </xacro:unless>
          <xacro:if value="${joint_limited}">
              <limit lower="${link_2_lower_limit}" upper="${link_2_upper_limit}" effort="20" velocity="2.7925" />
          </xacro:if>
        </joint>
        <!-- link3-->
        <link  name="link_3">
          <inertial>
            <origin xyz="0.021264 -0.0041182 0.082782"
                    rpy="0 0 0" />
            <mass value="${link_3_mass}" />
            <inertia ixx="0" ixy="0" ixz="0"
                      iyy="0" iyz="0" izz="0" />
          </inertial>
      
          <visual>
            <geometry>
              <mesh filename="package://multibot_efort/meshes/er10/link_3.STL" />
            </geometry>
            <material name="Orange" />
          </visual>
      
          <collision>
            <geometry>
              <mesh filename="package://multibot_efort/meshes/er10/link_3.STL" />
            </geometry>
          </collision>
        </link>
        
        <joint name="joint_3" type="revolute">
          <origin xyz="0 0 0.68"
                  rpy="0 0 0" />
          <parent link="link_2" />
          <child link="link_3" />
          <axis xyz="0 -1 0" />
          <xacro:unless value="${joint_limited}">
              <limit lower="${-2.0 * pi}" upper="${2.0 *pi}" effort="20" velocity="2.9671" />
          </xacro:unless>
          <xacro:if value="${joint_limited}">
              <limit lower="${link_3_lower_limit}" upper="${link_3_upper_limit}"  effort="20" velocity="3.1416" />
          </xacro:if>
        </joint>
        <!-- link4-->
        <link name="link_4">
          <inertial>
            <origin xyz="0.31875 0.0049527 0"
                    rpy="0 0 0" />
            <mass value="${link_4_mass}" />
            <inertia  ixx="0.0072478" ixy="0.0014639" ixz="0"
                      iyy="0.030284"  iyz="0" izz="0.037503" />
          </inertial>
      
          <visual>
            <geometry>
              <mesh
                filename="package://multibot_efort/meshes/er10/link_4.STL" />
            </geometry>
            <material name="Orange" />
          </visual>
      
          <collision>
            <geometry>
              <mesh
                filename="package://multibot_efort/meshes/er10/link_4.STL" />
            </geometry>
          </collision>
        </link>
        
        <joint name="joint_4" type="revolute">
          <origin xyz="0.2045 0 0.175"
                  rpy="0 0 0" />
          <parent link="link_3" />
          <child link="link_4" />
          <axis xyz="1 0 0" />
          <xacro:unless value="${joint_limited}">
              <limit lower="${-2.0 * pi}" upper="${2.0 *pi}" effort="20" velocity="2.9671" />
          </xacro:unless>
          <xacro:if value="${joint_limited}">
              <limit lower="${link_4_lower_limit}" upper="${link_4_upper_limit}"  effort="20" velocity="5.7596" />
          </xacro:if>
        </joint>
        <!-- link5-->
        <link name="link_5">
          <inertial>
            <origin xyz="0.018888 -0.0096333 -0.00018357"
                    rpy="0 0 0" />
            <mass value="${link_5_mass}" />
            <inertia ixx="0.00039655" ixy="-0.00015969" ixz="0"
                      iyy="0.00061483" iyz="0" izz="0.0009914" />
          </inertial>
          <visual>
            <geometry>
              <mesh  filename="package://multibot_efort/meshes/er10/link_5.STL" />
            </geometry>
            <material name="Orange" />
          </visual>
      
          <collision>
            <geometry>
              <mesh
                filename="package://multibot_efort/meshes/er10/link_5.STL" />
            </geometry>
          </collision>
        </link>
        
        <joint name="joint_5" type="revolute">
          <origin xyz="0.54 0 0"
                  rpy="0 0 0" />
          <parent link="link_4" />
          <child link="link_5" />
          <axis xyz="0 -1 0" />
          <xacro:unless value="${joint_limited}">
              <limit lower="${-2.0 * pi}" upper="${2.0 *pi}" effort="20" velocity="2.9671" />
          </xacro:unless>
          <xacro:if value="${joint_limited}">
              <limit lower="${link_5_lower_limit}" upper="${link_5_upper_limit}" effort="20" velocity="6.2832" />
          </xacro:if>
        </joint>
        <!-- link6-->
        <link name="link_6">
          <inertial>
            <origin xyz="0.010372 0 0"
                    rpy="0 0 0" />
            <mass value="${link_6_mass}" />
            <inertia ixx="0" ixy="0" ixz="0"
                      iyy="0" iyz="0" izz="0" />
          </inertial>
          <visual>
            <geometry>
              <mesh filename="package://multibot_efort/meshes/er10/link_6.STL" />
            </geometry>
            <material name="Orange" />
          </visual>
      
          <collision>
            <geometry>
              <mesh filename="package://multibot_efort/meshes/er10/link_6.STL" />
            </geometry>
          </collision>
        </link>
        
        <joint name="joint_6" type="revolute">
          <origin xyz="0.095 0 0"
                  rpy="0.000 0 0" />
          <parent link="link_5" />
          <child link="link_6" />
          <axis xyz="1 0 0" />
          <xacro:unless value="${joint_limited}">
              <limit lower="${-2.0 * pi}" upper="${2.0 *pi}" effort="20" velocity="2.9671" />
          </xacro:unless>
          <xacro:if value="${joint_limited}">
              <limit lower="${link_6_lower_limit}" upper="${link_6_upper_limit}" effort="10" velocity="10.472" />
          </xacro:if>         
        </joint>
    </xacro:macro>    
</robot>
